test/mpi/attr/fkeyvaltype.c

```diff
@@ -41,73 +42,87 @@ int delete_fn(MPI_Datatype type, int keyval, void *attribute_val, void *extra_st
 
 int main(int argc, char *argv[])
 {
-    int errs = 0;
+    int err, errs = 0;
     int attrval;
     int i, key[32], keyval, saveKeyval;
+    int type_idx, obj_idx;
     MPI_Datatype type, duptype;
-    MTestDatatype mstype, mrtype;
+    MPI_Datatype basic_type;
+    DTP_t dtp;
     char typename[MPI_MAX_OBJECT_NAME];
     int tnlen;
+    MPI_Comm comm = MPI_COMM_WORLD;
 
     MTest_Init(&argc, &argv);
 
-    while (MTestGetDatatypes(&mstype, &mrtype, 1)) {
-        type = mstype.datatype;
-        MPI_Type_create_keyval(copy_fn, delete_fn, &keyval, (void *) 0);
-        saveKeyval = keyval;    /* in case we need to free explicitly */
-        attrval = 1;
-        MPI_Type_set_attr(type, keyval, (void *) &attrval);
-        /* See MPI-1, 5.7.1.  Freeing the keyval does not remove it if it
-         * is in use in an attribute */
-        MPI_Type_free_keyval(&keyval);
+    for (type_idx = 0; type_idx < DTP_BASIC_TYPE_NUM; type_idx++) {
+        basic_type = DTP_basic_types[type_idx];
```

Comment:
I thought the intent was to run the test for just one basic datatype and not all of them.  Different basic datatypes would create different tests, isn't it?

Reply:
Yes that file hasn't changed yet. Please checkout the `test/mpi/pt2pt/sendrecv1.c`

Reply:
Yeah, the general idea seems correct.  The environment variable needs to be internal to DTPOOLS.

Reply:
Sure, will move it once review of the framework is done